core/arch/arm/kernel/mutex.c

```diff
@@ -60,6 +66,29 @@ static void __mutex_lock(struct mutex *m, const char *fname, int lineno)
 	}
 }
 
+static void __mutex_lock_recursive(struct recursive_mutex *m, const char *fname,
+				   int lineno)
+{
+	assert_have_no_spinlock();
+	assert(thread_get_id_may_fail() != THREAD_ID_INVALID);
+	assert(thread_is_in_normal_mode());
+
+	/* Thread ID has to fit in a short int */
+	COMPILE_TIME_ASSERT(CFG_NUM_THREADS <= SHRT_MAX);
+
+	if (atomic_load_short(&m->owner) == thread_get_id()) {
+		if (!refcount_inc(&m->lock_depth))
+			panic();
+		return;
+	}
+
+	__mutex_lock(&m->m, fname, lineno);
+
+	assert(m->owner == THREAD_ID_INVALID);
+	m->owner = thread_get_id();
```

Comment:
Wouldn't it make more sense to store this using `atomic_store_short()` to guarantee that all the bytes are written at the same time? It's very likely that happens in this direct assignment too, but without the guarantee.